extensions/markdown-language-features/src/features/smartSelect.ts

```diff
@@ -0,0 +1,238 @@
+						return parent;
+					} else {
+						return new vscode.SelectionRange(adjustedRange, parent);
+					}
+				} else if (parent.range.end.line === range.end.line) {
+					let adjustedRange = new vscode.Range(parent.range.start, range.end.translate(undefined, parent.range.end.character));
+					if (adjustedRange.isEqual(parent.range)) {
+						return parent;
+					} else {
+						return new vscode.SelectionRange(adjustedRange, parent.parent);
+					}
+				} else {
+					return parent;
+				}
+			} else {
+				return new vscode.SelectionRange(range);
+			}
+		}
+	} else {
+		return undefined;
+	}
+}
+
+function createFencedRange(token: Token, cursorLine: number, document: vscode.TextDocument, parent?: vscode.SelectionRange): vscode.SelectionRange {
+	const startLine = token.map[0];
+	const endLine = token.map[1] - 1;
+	let onFenceLine = cursorLine === startLine || cursorLine === endLine;
+	let fenceRange = new vscode.Range(new vscode.Position(startLine, 0), new vscode.Position(endLine, document.lineAt(endLine).text.length));
+	let contentRange = endLine - startLine > 2 && !onFenceLine ? new vscode.Range(new vscode.Position(startLine + 1, 0), new vscode.Position(endLine - 1, getEndCharacter(document, startLine + 1, endLine))) : undefined;
+	if (parent && contentRange) {
+		if (parent.range.contains(fenceRange) && !parent.range.isEqual(fenceRange)) {
+			return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(fenceRange, parent));
+		} else if (parent.range.isEqual(fenceRange)) {
+			return new vscode.SelectionRange(contentRange, parent);
+		} else if (rangeLinesEqual(fenceRange, parent.range)) {
+			let revisedRange = fenceRange.end.character > parent.range.end.character ? fenceRange : parent.range;
+			return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(revisedRange, getRealParent(parent, revisedRange)));
+		} else if (parent.range.end.line === fenceRange.end.line) {
+			parent.range.end.translate(undefined, fenceRange.end.character);
+			return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(fenceRange, parent));
+		}
+	} else if (contentRange) {
+		return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(fenceRange));
+	} else if (parent) {
+		if (parent.range.contains(fenceRange) && !parent.range.isEqual(fenceRange)) {
+			return new vscode.SelectionRange(fenceRange, parent);
+		} else if (parent.range.isEqual(fenceRange)) {
+			return parent;
+		} else if (rangeLinesEqual(fenceRange, parent.range)) {
+			let revisedRange = fenceRange.end.character > parent.range.end.character ? fenceRange : parent.range;
+			return new vscode.SelectionRange(revisedRange, parent.parent);
+		} else if (parent.range.end.line === fenceRange.end.line) {
+			parent.range.end.translate(undefined, fenceRange.end.character);
+			return new vscode.SelectionRange(fenceRange, parent);
+		}
+	}
+	return new vscode.SelectionRange(fenceRange, parent);
+}
+
+function isList(type: string): boolean {
```

Comment:
Consider having this take a token instead of a type string. That would make it more clear how this function is supposed to be used